React-hook-forms refactor of DynamicForm

ZachCMP opened the pull request:
## Description

Refactors the DynamicForm component to use `react-hook-form` and does some splitting up of components to make composing things easier. Couple notes:

1. This is currently draft, there is some cleanup left to do before I'd merge this, but I wanted to put it up for QA/reactions before proceeding with that
2. I've replaced all DynamicForm component with the refactored one by just making `DynamicForm` export the `RefactorForm` instead of the normal dynamic form. If there are instances in the app where components are using sub-components of DynamicForm or some such, those might not be handled right.
3. Due to the way that `react-hook-form` handles field names, the format `4.12.4` is troublesome, since it interprets that as a path (i.e. `{ '4': { '12': { '4': value } } }`). I've decided to deal with that by converting the periods to commas ('safe' format), and storing the values that way, converting them back to periods ('clean' format) where needed. This isn't ideal, but I don't see a way around it unless we can configure RHF to use a different delimiter for pathing. Using periods is typically the convention for path flattening in every case I've seen.
4. My immediate goal with this work was primarily to port over the current functionality, so there's definitely more that could be done in terms of enabling composition and handling certain workflows, as well as more work that could be done on refactoring the form logic code generally beyond what I've done here.

## Type of change
[//]: # 'remove options that are not relevant'
- [x] Code clean-up / housekeeping

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (eslint)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ZachCMP committed:
Initial spike work

ZachCMP committed:
Merge remote-tracking branch 'origin/release-99' into 186567726-forms-refactor-spike

ZachCMP committed:
Work RHF into all form mechanics

ZachCMP committed:
Pull form components apart, drill props

ZachCMP committed:
Cleanup form

ZachCMP committed:
WIP on dirty observer

ZachCMP committed:
WIP nearly feature complete

ZachCMP committed:
Feature complete, ready to start QA

ZachCMP committed:
Merge remote-tracking branch 'origin/release-100' into 186801632-rhf-refactor

ttoomey commented on src/modules/form/components/refactor/useFormDefinitionHandlers.tsx:
```diff
@@ -0,0 +1,186 @@
+  errors?: ValidationError[];
+}
+
+export const getSafeLinkId = (linkId: string) =>
+  String(linkId).replace(/\./g, ',');
+export const getCleanedLinkId = (linkId: string) =>
+  String(linkId).replace(/,/g, '.');
```
This seems like a reasonable work-around but it's too bad we need it. I wonder if we could change the link-ids in form definitions so that they no longer include dots. That would mean either we translate the fields in the form definition in graphql or perhaps a db migration.

Could we use some other replacement character? A Link id of `1,2,3` suggests that it's CSV formatted data rather than an id. Maybe use kebab-case `-` or double underscore `__` in place of dots?

ZachCMP replied:
I think the dots correspond to HUD spec, so I'm unsure how easy/okay it would be to replace them with another character at the backend or API level. I agree with the sentiment that the comma isn't ideal, and I'm open to the option of another replacement character. Though I think using any character that could possibly be part of a logical ID isn't the best, since it's possible it would be used in link IDs innocently and then cause issues or need special handling. For that reason I don't think using `-` or `_` is a good idea. My logic for using a comma is: 1) It reads as related to a period, so it's maybe more clear what character it's escaping; and 2) I would never think to use a comma in an ID, so I can't see one being used in a link ID in a future form definition.

In terms of picking a replacement for a comma, I see a couple of options:

1. Use some other character that wouldn't typically be included in IDs, maybe: `;`, `:`, `|`, `^`, `~`, `%`, etc.
2. Use an escape sequence of some kind: `:dot:` or some such.
3. Hash or base64 the link IDs. This would do it, and would probably make this a non-issue for any other reserved character issues, but probably not ideal since the internal form state would be hard to read.

What do you think?

ttoomey replied:
> For that reason I don't think using - or _ is a good idea

I imagine double or triple underscore would be safe (we'd leave single underscores alone and transform with `s/\./__/ and s/__/\./`).

Here's another idea:
These are HUD defined values so maybe we prefix them all with 'hud'? That would have the extra benefit of allowing us to target these properties with selectors (currently these ids that start with numeric chars are hard to use in js/css).

I'm thinking something like `hud-1-2-3` would be ideal, it looks nice and is not too hard to transform back to the original value with something like `/^hud-(\d+)-(\d+)-(\d+)/`. The 'hud' prefix should make it safe to use simple hyphens or underscores in place of dots.

ZachCMP replied:
I changed it to `__`

gigxz replied:
https://www.pivotaltracker.com/story/show/186922374

ttoomey commented on src/modules/form/components/refactor/RefactorField.tsx:
```diff
@@ -0,0 +1,482 @@
+  );
+};
+
+const MAX_INPUT_AND_LABEL_WIDTH = 600; // allow label to extend past input before wrapping
+export const MAX_INPUT_WIDTH = 500;
+const FIXED_WIDTH_MEDIUM = 350;
+const FIXED_WIDTH_SMALL = 200;
+const FIXED_WIDTH_X_SMALL = 100;
```
Looks like this component has quite a few responsibilities
* renders the correct input component for the item 
* determines the width of the component
* fetches pick-list options 
* sets up event handlers 
* validation / warnings (warn if empty, etc)

I know we decided we didn't want to touch the input components themselves for this round of work, so perhaps this is all we can do with that constraint. I'm fine with that; maybe we can plan come back for another round of refactoring when we have more space.

For future consideration, I'd suggest creating wrapper components for each input type that can manage these responsibilities.

ZachCMP replied:
Yeah, this is definitely one that I want to revisit in a future iteration. I consider this out of scope for now as I see this pass as primarily trying to rewire state and mechanics to use RHF. I've done some splitting out of components to that end, but I'd like to tackle these deeper input component pieces as their own piece of work.

ttoomey commented on src/modules/form/components/refactor/RefactorField.tsx:
```diff
@@ -0,0 +1,482 @@
+
+    default:
+      console.warn('Unrecognized item type:', item.type);
+      return <></>;
```
suggest throwing an exception in these cases

ttoomey commented on src/modules/form/components/refactor/RefactorFormLayout.tsx:
```diff
@@ -0,0 +1,26 @@
+    <>
+      <Grid container direction='column' spacing={2}>
+        {errors && <Grid item>{errors}</Grid>}
+        {children}
```
Small thing but I wonder if we can get away from the pattern of components that require children to be wrapped in `<Grid item>`. Maybe it should be this component's responsibility to wrap the children in Grid item?

```jsx
    <>
      <Grid container direction='column' spacing={2}>
        {errors && <Grid item>{errors}</Grid>}
        {React.Children.map(children, (child) => (<Grid item>{child}</Grid>))}
      </Grid>
      {saveButtons}
    </>
```

ZachCMP replied:
Agreed, I think the parents should be responsible to wrapping the children as much as possible. However, in this case the "children" is actually just the single form layout component, so I don't think we want to do this here.

ttoomey commented on package.json:
```diff
@@ -98,10 +98,12 @@
     "react-imask": "^7.1.3",
     "react-pdf": "^6.2.2",
     "react-router-dom": "^6.8.0",
+    "recoil": "^0.7.7",
```
looks like recoil didn't end up being needed

ZachCMP replied:
Yup, removed

ttoomey reviewed (commented):
Looking great! I left a few thoughts/ suggestions. 

There's some extra space under some input components now, probably dependent components getting rendered

![Screenshot_2024-01-22_at_12_50_36_PM](https://github.com/greenriver/hmis-frontend/assets/505569/9ee7c1ca-6f18-4e7c-a968-51658727ed78)

ZachCMP committed:
Merge remote-tracking branch 'origin/release-101' into 186801632-rhf-refactor

ZachCMP committed:
Fixup dependent wrappers

ZachCMP commented:
@ttoomey Addressed your feedback and pushed some updates. The gaps should mostly be gone now unless the group components are trying to wrap their children in some other component. This was another consequence of inverting the disabled/autofill control. It was rendering groups where all the children were disabled, so I added additional checks in the wrapper to hide the group if all of its children are also hidden.

gigxz commented on src/modules/form/data/service.json:
```diff
@@ -0,0 +1,588 @@
+{
```
Can you move it into the `test/__mocks__` dir since it is not used in the applciation

ZachCMP replied:
Used for a story so I could test different form defs. I'll probably remove most of the ones I added since I don't need them beyond development

gigxz commented on src/test/__mocks__/pickLists.ts:
```diff
@@ -0,0 +1,379 @@
+import { PickListOption } from '@/types/gqlTypes';
+
+export const pickListData: Record<string, PickListOption[]> = {
```
You should be able to pull these from the gqlEnums file

ZachCMP replied:
I actually don't think that I can. I need the `groupLabel` and `groupCode`, which I don't think are included in that file. But I'll probably end up removing this since I don't think I'll end up leaving it in storybook beyond the dev on this feature.

gigxz commented on yarn.lock:
```diff
@@ -10817,6 +10827,13 @@ recast@^0.23.1:
     source-map "~0.6.1"
     tslib "^2.0.1"
 
+recoil@^0.7.7:
```
I dont think this should be here, canyou run yarn and recommit

gigxz commented:
Conditional logic weirdness repro:
1. Open an intake assessment in a permanent housing project
2. Set Prior Living values:
<img width="739" alt="Screenshot 2024-01-26 at 10 19 08 AM" src="https://github.com/greenriver/hmis-frontend/assets/5333982/72040449-569e-44f7-a57c-5846a8753150">
3. Submit, re-open the assessment for editing
4. Clear out the Length of Stay value. There is an empty gray bubble:

<img width="649" alt="Screenshot 2024-01-26 at 10 18 03 AM" src="https://github.com/greenriver/hmis-frontend/assets/5333982/6d028a24-baef-47c5-abba-066c41a4b13f">



The previous implementation had logic for groups, where if all children in a group were hidden, then the group component would be hidden as well. I think we need to keep that logic. I think it's missing which is what is causing the above.

gigxz commented:
Disabling condition had some logic where the Overall Disabling Condition would be fixed to `Yes` and not editable if it was set from a system rule. I'd be happy to look at that later, I think it was internal to the disability table logic. So I wouldn't worry too much about it now, just want to flag it so it doesnt get forgotten.

ZachCMP committed:
yarn.lock

ZachCMP committed:
Fixup depedency wrapper to use all children recursively

ZachCMP commented:
Yeah, the group components need some additional work. They assume they'll be passed `values` but that's not really true any more, so I need to find a workaround. I could pull the RHF logic into them, which would probably be best performance, but I could also wrap them in a `Watcher` or something and just pass the values from there and leave them in the old pattern. I'll see what makes most sense when I get there.

ZachCMP commented:
@gigxz I pushed fixes for the conditional logic. Could you take a look when you get a sec and see if it fixes it for you? I expanded the amount of things that the dependent wrapper listens to and what values it considers.

ZachCMP committed:
Replace current dynamic form

ZachCMP committed:
Remove refactor dir; move hook

ZachCMP committed:
Rename everything back to dynamic*

gigxz commented:
Yep that fixed the issue on PLS, thanks!

ZachCMP committed:
Fixup form group behavior

ZachCMP committed:
Cleanup storybook bloat

ZachCMP committed:
Remove change

ZachCMP committed:
Remove change

ZachCMP committed:
Remove change

ZachCMP committed:
Fix types

ZachCMP committed:
Address circular deps

ZachCMP committed:
Remove useDynamicFields

ZachCMP commented:
@gigxz @ttoomey This should be pretty close to ready to merge, so I'm taking if off draft status. I replaced all of the existing DynamicForm components with the new set of components and tweaked the behavior of everything to work with the new paradigm. Broadly speaking, I've tried to keep the RHF-dependent stuff in the base components and leave `DynamicField` and `DynamicGroup` with the values and setter paradigm they had before so we can use them outside of a RHF context. The component doing most of the heavy lifting in that regard is `DynamicFormField`, which handles making the RHF mechanics work with the value/setter paradigm.

I also changed the exports of DynamicForm around so there's a version that takes `handlers`–the RHF-based mechanics returned from `useFormDefinitionHandlers`, and a version that provides its own handlers. I've made the one that provides its own handlers the default export since that's the current paradigm, but the other is available if you want to declare the handlers at a higher level and pass them into the form (i.e. for assessments and such). I've left the `useImperativeHandle` bits in, since I wasn't sure how those pieces were being used and I didn't want to break anything.

One thing to note is that this removes the `useDynamicFields` hook that implemented the old react state mechanics. `DynamicForm` was the primary consumer of that hook, but it was also used by `DynamicView` and `BulkAdd`. I reworked DynamicView to use the new mechanics instead, as well as the BulkAdd component, but it doesn't look like that latter is used anywhere at the moment. It's used by `BulkAddServices`, but that component doesn't seem to be imported by anything currently. Regardless, some extra QA around those components would be helpful to make sure they're working well too. A quick look in storybook seems to indicate that at least the DynamicView is working, but I don't know every place it's used in the app and I don't have a utility to see a viewable version of the forms in the JSON form editor.

Take a look and let me know if there's anything else you'd like to see on this.

ttoomey commented on src/modules/form/components/AutofillFormItemWrapper.tsx:
```diff
@@ -0,0 +1,62 @@
+  getDependentLinkIds,
+}) => {
+  const { autofillInvertedDependencyMap, getAutofillValueForField } = handlers;
+  const name = (
```
suggest rename to dependentLinkIds

ttoomey commented on src/modules/form/components/DependentFormItemWrapper.tsx:
```diff
@@ -0,0 +1,75 @@
+    [item, itemMap]
+  );
+
+  const name = useMemo(() => {
```
suggest rename to dependentLinkIds

ttoomey commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,276 @@
+  const severalItemsChanged: SeveralItemsChangedFn = useCallback(
+    ({ values, type }) => {
+      Object.entries(values).forEach(([linkId, value]) =>
+        handlers.methods.setValue(getSafeLinkId(linkId), value, {
+          shouldDirty: type === ChangeType.User,
+        })
```
Maybe dry this up and use itemChanged so we get parseInt/Float?
```suggestion
        itemChanged({linkId, value, type})
```

ttoomey commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,276 @@
+  renderFn?: (children: ReactNode) => ReactNode;
+}
+
+const DynamicFormField: React.FC<Props> = ({
```
My understanding is the stack here is something like this. Does that seem right?
```js
DynamicFormField(renderFn)(
  // renderItemWithWrappers
  DependentFormItemWrapper(
    AutofillFormItemWrapper(
      //itemGroup, table etc
      (item) => item.children.map (
        ValueWrapper(DynamicField)
      )
    )
  )
)
```

ZachCMP replied:
@ttoomey Yeah, that's right. These are the responsibilities:

- `DynamicFormField` handles wiring up the form item to the form state handlers. It will use `renderItemWithWrappers` to wrap the children. It will only apply the wrappers that are appropriate, so it won't wrap a component with an autofill wrapper if the form item has no autofill logic.
- `DependentFormItemWrapper` handles hiding/showing the component based on the `enableWhen` logic. It does that by watching anything that the component depends on and updating the rendering when they change.
- `AutofillFormItemWrapper` does the same thing for autofill logic, and it handles updating the form values with the autofilled values when they are autofilled. It also uses `dirty` to only autofill if the field is not dirty, and does not mark the field dirty when it autofills.
- `ValueWrapper` pulls `value` out of the form state, allowing us to keep the value/setter pattern for the actual `DynamicField` and `DynamicGroup` components. So this lets us keep that pattern but isolate the re-renders per-field.

ttoomey replied:
Thanks for the write up! The description is helpful to get oriented in the form mechanics. Maybe the above could become inline comments/docs? (if it's not already there)

ttoomey reviewed (approved):
Looks good to me!! Great work 🙌

ZachCMP committed:
PR feedback

ZachCMP committed:
Merge remote-tracking branch 'origin/release-103' into 186801632-rhf-refactor

ZachCMP commented:
@ttoomey Addressed your feedback and pushed

ZachCMP commented:
@gigxz Anything else you need from me on this?

gigxz committed:
Merge branch 'release-103' into 186801632-rhf-refactor

gigxz committed:
Merge branch 'release-111' into 186801632-rhf-refactor

gigxz committed:
Additional conflicts and fix warnIfEmpty

gigxz committed:
Additional conflicts and fix warnIfEmpty

gigxz committed:
Merge branch 'release-115' into 186801632-rhf-refactor

gigxz committed:
Merge branch 'release-116' into 186801632-rhf-refactor

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,272 @@
+          {...fieldProps}
+        />
+      ) : (
+        <ValueWrapper name={getSafeLinkId(item.linkId)} handlers={handlers}>
+          {(value) => (
+            <DynamicField
+              value={value}
+              key={item.linkId}
+              item={item}
+              itemChanged={itemChanged}
```
Note to self:

ValueWrapper provides value for this form item from RHF state; `itemChanged` uses `setValue` internally to update the form state. This lets us use all the existing input components and `DynamicField` without modification.

(As opposed to adjusting DynamicField, or the input components themselves, to consume the RHF context)

ttoomey committed:
Merge branch 'release-146' into 186801632-rhf-refactor

ttoomey committed:
Merge branch 'release-147' into 186801632-rhf-refactor

ttoomey committed:
remove unused code

ttoomey committed:
misc fixes

ttoomey committed:
186801632-rhf-refactor

second attempt at fixing MCI

ttoomey committed:
fix console warning about key

ttoomey committed:
fix illegal dom nesting

ttoomey committed:
include hidden fields for autocomplete

ttoomey committed:
comment

ttoomey committed:
fix missing errors

ttoomey committed:
Merge branch 'release-148' into 186801632-rhf-refactor

ttoomey commented on src/modules/errors/components/ApolloErrorTrace.tsx:
```diff
@@ -25,8 +25,9 @@ const ApolloErrorTrace: React.FC<Props> = ({ errors }) => {
         <CommonUnstyledList key={JSON.stringify(e)}>
-          {(e?.backtrace || []).map((line) => (
-            <li>{line}</li>
+          {(e?.backtrace || []).map((line, idx) => (
```
fixes warning about keys (trying to reduce chatter in the console)

gigxz replied:
thanks!

ttoomey commented on yarn.lock:
```diff
@@ -8139,10 +8139,10 @@ react-dropzone@^14.2.3:
-  version "7.52.1"
-  resolved "https://registry.yarnpkg.com/react-hook-form/-/react-hook-form-7.52.1.tgz#ec2c96437b977f8b89ae2d541a70736c66284852"
-  integrity sha512-uNKIhaoICJ5KQALYZ4TOaOLElyM+xipord+Ha3crEFhTntdLvWZqVY49Wqd/0GiVCA/f9NjemLeiNPjG7Hpurg==
+react-hook-form@^7.54.2:
```
this is optional, seemed like this release was just bug fixes and non-breaking changes so worth doing

ttoomey commented on src/modules/hmis/components/HmisEnum.tsx:
```diff
@@ -29,7 +29,7 @@ interface Props extends TypographyProps {
   const [label, color] = getLabelAndColor(enumMap, value);
   return (
-    <Typography variant='body2' color={color} {...props}>
+    <Typography variant='body2' color={color} component='span' {...props}>
```
fixes invalid dom nesting warning

ttoomey commented:
@gigxz I've got this merged and the tests are passing. I'm going to take a another pass at reviewing it

ttoomey commented on src/modules/external/mci/components/MciClearance.tsx:
```diff
@@ -296,23 +296,39 @@ const MciClearanceWrapper = ({
+  const { getValues } = methods;
+
+  // watch specific form fields.
+  // FIXME: I suspect that this should translate MCI_CLEARANCE_FIELDS into form fields instead of hard-coding them
```
This seems to work but I should probably clean this up a bit.

ttoomey commented on src/modules/form/components/client/addresses/SimpleAddressInput.tsx:
```diff
@@ -27,7 +27,11 @@ const SimpleAddressInput: React.FC<Props> = ({
-      onChange([createInitialValue()]);
+      // setTimeout is needed because if the inputs were not previously rendered, they are not yet registered. So the
+      // timeout allows them to register before we change the value here
+      setTimeout(() => {
+        onChange([createInitialValue()]);
+      });
```
I think the other way to fix this would be to descend into all the inputs for hidden groups and register them when the form is displayed. Maybe that would be a more robust solution? I think otherwise this problem would occur anytime we show fields and write initial values in a useEffect

ttoomey replied:
doesn't actually appear to be registration related. Not totally sure why this doesn't work but dropping for a cleaner defaultValue solution

ttoomey committed:
additional dom-nesting fix

ttoomey committed:
cleanup mci

ttoomey committed:
cleanups around form actions and layout

ttoomey commented on src/modules/form/hooks/useFormDefinitionHandlers.tsx:
```diff
@@ -0,0 +1,213 @@
+  viewOnly?: boolean;
+}
+
+export const getSafeLinkId = (linkId: string) =>
+  String(linkId).replace(/\./g, '__');
+export const getCleanedLinkId = (linkId: string) =>
+  String(linkId).replace(/__/g, '.');
```
I don't think these are needed anymore as we have cleaned up our link id format. I guess they aren't hurting anything but probably safe to remove(?)

gigxz replied:
Yes, I think it's a good idea to remove them. All the forms have been updated to have a valid format

ttoomey replied:
Will do, I'll leave them for the moment; removing them may cause some code-churn; I want to address the more immediate bugs first

ttoomey replied:
I've removed them now

ttoomey commented on src/modules/form/hooks/useFormDefinitionHandlers.tsx:
```diff
@@ -0,0 +1,213 @@
+        values,
+        itemMap,
+        localConstants,
+        viewOnly, // TODO: Handle this
```
I'm not sure what needs to be done here

gigxz replied:
It appears handled to me.
Related PR https://github.com/greenriver/hmis-frontend/pull/588/commits/5c7955c248dde71053de0c60728f4a881fb1afc0 added support for `autofill_readonly` attribute to specify whether the field should autofill even when form is being viewed as "read only".

One such example is Monthly Total Income. To reproduce, I did:
* Fill out and submit assessment
* In the rails console, manually change MonthlyTotalIncome to a different value (aka make it so incomes don't add up correctly)
* Reopen the submitted assessment. It correctly shows the "wrong" value ($400) rather than the auto-calculated value. Screenshot below. This is what we expect. (This comes up for data correction, it's beneficial to see the **real** stored value when viewing. Re-submitting the form will fix the calculation).

<img width="833" alt="Screenshot 2025-01-13 at 11 14 38 AM" src="https://github.com/user-attachments/assets/5cfd2b2e-3c86-4afc-b7fe-3db66a557eda" />

ttoomey replied:
thanks, I've removed the TODO comment

ttoomey committed:
tidy autocomplete and dependent wrappers

ttoomey committed:
comments

gigxz commented on src/modules/form/components/AutofillFormItemWrapper.tsx:
```diff
@@ -0,0 +1,60 @@
+  //   link_id: item.id,
+  //   autofill_values: [ { autofill_when: [{ question: dependentLinkIds[0] }] } ],
+  const { autofillInvertedDependencyMap, getAutofillValueForField } = handlers;
+  const dependentLinkIds =
+    autofillInvertedDependencyMap[item.linkId].map(getSafeLinkId);
```
```suggestion
  const dependentLinkIds = (
    autofillInvertedDependencyMap[item.linkId] || []
  ).map(getSafeLinkId);
```


This fixes the following bug. I haven't dug into it too much beyond this. Why did we think that the `item.linkId` would definitely be present in `autofillInvertedDependencyMap`?

Repro:
* Add a new member to a household. They should have incomplete status.
* Fill out intake assessment for new member, hit "save"
* Go to the summary/submission tab in the household workflow
* Submit the intake for the new member
* Navigate back to the intake for the new member

Without this fix I was seeing an error:
<img width="1671" alt="Screenshot 2025-01-13 at 11 11 30 AM" src="https://github.com/user-attachments/assets/ff4dd2a7-59a3-4a62-b448-878b9284e828" />

ttoomey replied:
Thanks! I totally forgot to check the household enrollment form. I'll fix it up

gigxz replied:
Thanks!
@ttoomey I think I misrepresented this a bit. I am seeing this error when viewing most **submitted** assessment (tried: a local custom assessment, a hud intake, a hud annual), regardless of whether it's in the individual or household view.

The Intake Form Definition I'm using is the one that gets seeded by default in a development environment and for `CLIENT=qa_hmis` -- that's the one that has _all_ the HUD questions enabled. Maybe you're using something different? 

Assessments that I _can_ open successfully:
* Any work-in-progress assessments (HUD and custom)
* A tiny submitted "housing needs assessment" I have locally -- the difference may be that it doesn't have conditionals or autofills (Not sure, but I can track down the differences if you have trouble reproducing)

ttoomey replied:
Yes, I was testing with definitions from a different client env to exercise the MCI component, thanks for the reminder on using QA definitions. Your fix to autofillInvertedDependencyMap seems to corect the crash

gigxz commented on src/modules/form/util/formUtil.tsx:
```diff
@@ -1375,11 +1388,12 @@ export const getDependentItemsDisabledStatus = ({
 }: GetDependentItemsDisabledStatus) => {
```
Poking around at this because I'm looking at a difference with the DisabilityTable. Maybe I'm missing something obvious but I'm not seeing that `getDependentItemsDisabledStatus` is used anymore? (Or maybe it already wasn't?)

ttoomey replied:
good catch! It was used previously but I've removed it

gigxz commented on src/modules/form/components/group/DisabilityTable.tsx:
```diff
@@ -52,7 +52,7 @@ const DisabilityTable = ({
       ?.linkId;
   }, [item]);
 
```
Not sure where this comment will belong, but placing it here. There is a difference in something between QA and this branch:

* Select Yes to Physical Disability
* Select Yes to Physical Disability is a Disabling Condition
* Clear out the value for Physical Disability
  * **QA**: this results in "Physical Disability is a Disabling Condition" being cleared out as well
  * **This Branch**:  "Physical Disability is a Disabling Condition" appears to remain as Yes.

Pretty sure this doesn't stem from logic within the DisabilityTable component (which mostly deals with the "overall disabling condition" at the bottom), but rather something about "autofill dependencies" . Let me know if you'd like me to spend some more time tracking it down or pairing!

ttoomey replied:
good find! I've patched this behavior in ~~DependentFormItemWrapper~~ DynamicFormField (d2a32a047d3c1a715b833b2fc1f20931d8eb757c)

ttoomey committed:
bugfixes

gigxz commented on src/modules/form/components/DependentFormItemWrapper.tsx:
```diff
@@ -0,0 +1,118 @@
+        isItemDisabled(child) &&
+        child.disabledDisplay === DisabledDisplay.Hidden
+    );
+    return [childrenHidden, dependantValues];
```
I'm confused at the return of `dependantValues` here which doesn't seem to be used?

ttoomey replied:
The intention was to ensure isDisabled is re-evaluated when dependantValues change; however including in the dependency array without referencing it from the function is a linter error. Returning the dependency lets us include it without needing magic comments. 

But this isn't super clear; I could:
1) add a comment
2) remove this memoization (maybe we can rely on isolated renders to keep this from being called too frequently)
3) rewrite it to be clearer using useState/useEffect

gigxz replied:
Ah I see, thanks. I think adding a comment would be great

gigxz commented on src/modules/form/components/DependentFormItemWrapper.tsx:
```diff
@@ -0,0 +1,118 @@
+    dependantValues,
+  ]);
+
+  // If the element has become disabled, clear the value
+  const { getValues, setValue } = handlers.methods;
+  useEffect(() => {
+    if (!isDisabled) return;
+    if (isNil(getValues(linkId))) return;
+
+    setValue(linkId, null, { shouldDirty: false });
+  }, [isDisabled, getValues, linkId, setValue]);
```
Suggestion to not clear value if item has `DisabledDisplay.ProtectedWithValue`. This matches current behavior. (Tested)
```suggestion
  // If the element has become disabled, clear the value (if appropriate)
  const { getValues, setValue } = handlers.methods;
  useEffect(() => {
    if (!isDisabled) return;
    // If item is configured with "ProtectedWithValue", the value should remain even when the item is disabled.
    if (item.disabledDisplay === DisabledDisplay.ProtectedWithValue) return;
    if (isNil(getValues(linkId))) return;

    setValue(linkId, null, { shouldDirty: false });
  }, [isDisabled, getValues, linkId, setValue, item.disabledDisplay]);
```

Desired:
![Kapture 2025-01-14 at 09 41 41](https://github.com/user-attachments/assets/72022e69-f49f-4c62-ae15-92042c69f66d)

Bug:
![Kapture 2025-01-14 at 09 43 53](https://github.com/user-attachments/assets/e1c031d5-af1f-4f51-a73a-a9d4c888f1a1)


---


The "disabled display" functionality is not documented in the API, but it is described in the form builder:

<img width="539" alt="Screenshot 2025-01-14 at 9 39 19 AM" src="https://github.com/user-attachments/assets/de5d4bc1-dcc9-4fc9-808d-83e45dc4d1b5" />



Form items used in gif example:
```
        {
          "text": "test2",
          "type": "STRING",
          "bounds": [
            {
              "id": "bound_0b0fc13f",
              "type": "MAX",
              "severity": "error",
              "value_number": 6
            },
            {
              "id": "bound_962ccc5e",
              "type": "MIN",
              "severity": "error",
              "value_number": 3
            }
          ],
          "link_id": "test_2",
          "mapping": {
            "custom_field_key": "customassessment_test_2"
          },
          "required": false,
          "read_only": false,
          "helper_text": "min/max",
          "warn_if_empty": false,
          "disabled_display": "HIDDEN"
        },
        {
          "text": "number enabled when previous text is present",
          "type": "INTEGER",
          "link_id": "number_2",
          "mapping": {
            "custom_field_key": "customassessment_number_2"
          },
          "required": false,
          "read_only": false,
          "enable_when": [
            {
              "operator": "EXISTS",
              "question": "test_2",
              "answer_boolean": true
            }
          ],
          "helper_text": "disabled display = protected with value",
          "warn_if_empty": false,
          "enable_behavior": "ALL",
          "disabled_display": "PROTECTED_WITH_VALUE"
        },
```

ttoomey replied:
Thanks, that was very helpful. I think I've located where the original branch was trying to handle this case (src/modules/form/components/DynamicField.tsx:97). But that doesn't seem to work due to props-spreading overrides.

Perhaps I should back out my fix and fix this bug instead? This would correct the display without actually nulling-out the form value; I'm a little unsure which is better
 
```diff
--- a/src/modules/form/components/DynamicFormField.tsx
+++ b/src/modules/form/components/DynamicFormField.tsx
@@ -1,5 +1,5 @@
 import { Grid } from '@mui/material';
-import { isEmpty, pick } from 'lodash-es';
+import { isEmpty, omit, pick } from 'lodash-es';
 import React, { cloneElement, ReactNode, useCallback } from 'react';
 
 import {
@@ -240,14 +240,14 @@ const DynamicFormField: React.FC<Props> = ({
               key={item.linkId}
               item={item}
               itemChanged={itemChanged}
-              disabled={isDisabled}
+              disabled={isDisabled || fieldProps?.disabled}
               errors={getFieldErrors(item)}
               horizontal={horizontal}
               pickListArgs={pickListArgs}
               warnIfEmpty={warnIfEmpty}
               // Needed to support referencing local constants in expression evaluation (DynamicDisplay)
               localConstants={localConstants}
-              {...fieldProps}
+              {...omit(fieldProps, ['disabled'])}
               inputProps={{
                 ...fieldProps?.inputProps,
                 ...buildCommonInputProps({
```

ttoomey committed:
second fix for clearing disabled fields

ttoomey committed:
register all fields on form init

ttoomey committed:
improvements

* call useForm with defaultValues instead of live-updating values to
reduce potential sources of state change
* provide a generic helper to walk the form definition

ttoomey committed:
bugfix rhf api use when observing dirty state for one field

ttoomey committed:
improve default address handling

ttoomey commented on src/modules/form/components/AutofillFormItemWrapper.tsx:
```diff
@@ -0,0 +1,61 @@
+  const autofillValue = getAutofillValueForField(item);
+
+  // Listen to see if this field has been edited by the user
+  const { isDirty } = useFormState({ control: control, name: linkId });
```
there was a bug here; isDirty is not the way to know if a single form field is diry

ttoomey committed:
cleanup

* remove explicit registration
* add comment

ttoomey committed:
Merge branch 'release-148' into 186801632-rhf-refactor

ttoomey committed:
remove getSafeLinkId and getCleanedLinkId

gigxz commented on src/modules/form/components/AutofillFormItemWrapper.tsx:
```diff
@@ -43,10 +43,12 @@ const AutofillFormItemWrapper: React.FC<Props> = ({
   useEffect(() => {
     // Don't autofill this field if it's been edited (i.e. is dirty)
+    // for example, we automatically set the radio choice for Income Source to true if any of the income fields are non-zero
     if (isDirty) return;
```
Note-to-self: do we need to differentiate between user-modified dirty fields vs system-modified (initial values, autofills)? see if behavior matches in those cases

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,292 @@
+            debug={
+              import.meta.env.MODE === 'development'
+                ? (keys?: string[]) => {
+                    const sectionValues = keys ? pick(values, keys) : values;
```
An adjustment I made locally so I could use the "Debug" button on the form during local development. `values` is stale here otherwise
```suggestion
                    const currentValues = handlers.getValues();
                    const sectionValues = keys
                      ? pick(currentValues, keys)
                      : currentValues;
```

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,292 @@
+              // Needed to support referencing local constants in expression evaluation (DynamicDisplay)
+              localConstants={localConstants}
+              {...omit(fieldProps, ['disabled'])}
+              inputProps={{
+                ...fieldProps?.inputProps,
+                ...buildCommonInputProps({
+                  item,
+                  values,
+                  localConstants: localConstants || {},
+                }),
```
There is a sneaky bug here. The reason `buildCommonInputProps` accepts values is because it calculates the `min` and `max` props to pass, based on the `bound`(s) specified on the item. Those min/max values can be based on other questions that have changing values. That functionality breaks here since we pass the stale `values` rather than `handler.getValues()`.


There is probably a way to optimize the renders a bit more, but, this does fix the functionality for me:

```suggestion
              inputProps={{
                ...fieldProps?.inputProps,
                ...buildCommonInputProps({
                  item,
                  values: handlers.getValues(),
                  localConstants: localConstants || {},
                }),
```


Example:
![Kapture 2025-01-16 at 09 46 44](https://github.com/user-attachments/assets/5fdf179e-4ca2-4735-9e28-4b3dff42fc8a)

```json
        {
          "text": "Q1",
          "type": "INTEGER",
          "link_id": "number_2",
          "mapping": {
            "custom_field_key": "customassessment_number_2"
          },
          "required": false,
          "read_only": false,
          "enable_when": [
            {
              "operator": "EXISTS",
              "question": "test_2",
              "answer_boolean": true
            }
          ],
          "helper_text": "disabled display = protected with value",
          "warn_if_empty": false,
          "enable_behavior": "ALL",
          "disabled_display": "PROTECTED_WITH_VALUE"
        },
        {
          "text": "Q2",
          "type": "INTEGER",
          "bounds": [
            {
              "id": "bound_95078fbc",
              "type": "MAX",
              "question": "number_2",
              "severity": "error"
            }
          ],
          "link_id": "q_2",
          "mapping": {
            "custom_field_key": "customassessment_q_2"
          },
          "helper_text": "max = value of Q1",
          "disabled_display": "HIDDEN"
        },
```

ttoomey replied:
should be fixed now, I improved error tracking in NumericInput

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,292 @@
+        <Grid item key={item.linkId}>
+          <DynamicViewField
+            item={item}
+            value={values[item.linkId]}
```
Assumption here that `values` is unchanging, but we do pass that `itemChanged` below for some specific use case ... one of those tiny complicated things we have added over time. Note-to-self to come back and test that scenario, to see if behavior regressed due to "stale" values being passed now.

gigxz replied:
Alright yeah, we do need the value wrapper here, to make behavior consistent with the main branch. It would also make this consistent with `DynamicViewFields` which also wraps `DynamicViewField` elements in a ValueWrapper to get the "actual" value.

Example:
* Q1 is a numeric field
* Q2 is a numeric field marked as "read only", with an autofill rule stating that it should autofill to value `500` if the value of Q1 is greater than 5. (Found a bug in the form builder when trying to configure this btw,  https://github.com/greenriver/hmis-frontend/pull/1027)

<img width="222" alt="Screenshot 2025-01-16 at 5 41 24 PM" src="https://github.com/user-attachments/assets/dca941f3-2b7a-4690-b0fd-b590a4c1ee08" />

--

Suggestion that fixes it (tested):

```diff
+           // Even though this field is marked as "read only", it needs the current value because it could change via autofill, or through the `adjustValue` fn which may change a value to have a readable label based on a remote pick list
+          <ValueWrapper name={getSafeLinkId(item.linkId)} handlers={handlers}>
+            {(value) => (
              <DynamicViewField
                item={item}
+                value={value}
                disabled={isDisabled}
                horizontal={horizontal}
                pickListArgs={pickListArgs}
                // Needed because there are some enable/disabled and autofill dependencies that depend on PickListOption.labels that are fetched (PriorLivingSituation is an example)
                adjustValue={itemChanged}
                // Needed to support referencing local constants in expression evaluation (DynamicDisplay)
                localConstants={localConstants}
                {...fieldProps}
              />
+            )}
+          </ValueWrapper>
```

ttoomey replied:
should be fixed now

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,292 @@
+  props: fieldProps,
+  renderFn,
+}) => {
+  const values = handlers.getValues();
```
Can you add a comment about how this `values` contains the form values as they were when this component was last rendered, but does not necessarily represent the _current_ form state (since there may have been other input changes that didn't trigger a re-render of this component). Unlike "use watch" . (If I got that right). Something I know will/has tripped me up :)

ttoomey replied:
I've adjust the implementation to use RHF useWatch instead

gigxz commented on src/modules/form/components/DynamicFormField.tsx:
```diff
@@ -0,0 +1,292 @@
+                </AutofillFormItemWrapper>
+              );
+            }}
+            values={values}
```
Note to self: look more closely at this interaction with `FormCard` and any other DynamicGroup that looks at `values` prop. We already have the special case for DisabilityTable below that passes the current values.

Maybe this could be simplified by passing the group the "values" of just the child link ids within that group. Or have the group watch what it needs separately via context.

gigxz replied:
adjustment proposed https://github.com/greenriver/hmis-frontend/pull/1032

gigxz reviewed (commented):
This is looking great. Leaving a batch of in-progress review comments- several of these are places I plan to come back to (hopefully today) to test/QA more closely.

ttoomey committed:
Update src/modules/form/components/DynamicFormField.tsx

Co-authored-by: Gig <gig@greenriver.org>

ttoomey committed:
Merge branch '186801632-rhf-refactor' of github.com:greenriver/hmis-frontend into 186801632-rhf-refactor

ttoomey committed:
cleanup dependency wrapper, support changing bounds

update validation error when bounds change

ttoomey committed:
value wrapper and view fixes

ttoomey committed:
hook improvement

ttoomey committed:
cleanups and bug fixing

ttoomey commented:
@gigxz - per our conversation on Friday, I did some additional refactoring to try and squash bugs and simplify form interactions.

gigxz commented:
Awesome thanks @ttoomey  I'll do another QA pass!

ttoomey commented on src/modules/form/hooks/useDynamicFormContext.tsx:
```diff
@@ -1,15 +1,21 @@
 };
 
-export const DynamicFormContext = createContext<DynamicFormContextType>({});
+export const DynamicFormContext = createContext<DynamicFormContextType>(
```
This context makes it convenient to access static form configuration. I find the context makes it easier to make smaller composable hooks since I don't need to pass around a bag of form props.

gigxz commented on src/modules/form/hooks/rhf/useDynamicFieldStatus.tsx:
```diff
@@ -0,0 +1,89 @@
+ *
+ * Attempts to isolate re-renders by watching only the items dependencies and those of the child items
+ *
+ */
```
This is much easier to understand and a helpful comment, thank you!

gigxz commented on src/modules/form/hooks/rhf/useDynamicFieldWatchValues.tsx:
```diff
@@ -0,0 +1,22 @@
+import { useMemo } from 'react';
+import { FieldValues, useFormContext, useWatch } from 'react-hook-form';
+
+export const useDynamicFieldWatchValues = (watchFields: string[]) => {
+  const { control } = useFormContext();
+  const valueArray = useWatch({
+    control,
+    name: watchFields,
+  });
```
```suggestion
export const useDynamicFieldWatchValues = (watchFields?: string[]) => {
  const { control } = useFormContext();
  const valueArray = useWatch({
    control,
    name: watchFields || [],
  });
```
I noticed a performance issue, and this fixes it. Passing `undefined` here causes fields that have no dependencies to watch ALL other fields.

Underlying type issue is that `LinkIdMap` doesn't allow for undefined value, but the `autofillInvertedDependencyMap` doesn't contain all link ids (or have non-undefined values for all link ids).

So either:
* Update map(s) to include all link ids mapped to `[]` when no dependencies (this would be an adjustment to buildAutofillDependencyMap I think, and maybe the inver logic?), or
* Add secondary type like `PartialLinkIdMap = Record<string, string[] | undefined>;` to use for autofill map which we know contains only partial set of keys

Second option seems simpler to me.

ttoomey replied:
Thanks for finding and fixing that; I agree that improving the types is probably simplest, I'll do that

ttoomey committed:
Update src/modules/form/hooks/rhf/useDynamicFieldWatchValues.tsx

Co-authored-by: Gig <gig@greenriver.org>

gigxz committed:
Unregister disabled fields, clean up DynamicGroup, fix household dirty state bug (#1032)

* Remove some dynamic group props
* Match previous behavior for DisabledDisplay.Protected which affects disability table
* Move console debug to hook
* Unregister disabled items
* Fix household dirty state bug
* Cleanup: remove unused isItemDisabled fn
* Comments, stop passing old getValues

gigxz committed:
Merge branch 'release-150' into 186801632-rhf-refactor

gigxz commented:
Rebased to 150 and opened https://github.com/greenriver/hmis-warehouse/pull/5097 to run system tests again

ttoomey committed:
186801632-rhf-refactor

* fix issue with unregistering inputs on unmount
* ensure fields are re-registered on mount
* safer type for invertDependencyMap

ttoomey committed:
186801632-rhf-refactor

* only render forms with the complete values needed for form logic to
  fix bugs

ttoomey commented:
@gigxz made a few fixes to address the QA issues (see thread  in [6043](https://github.com/open-path/Green-River/issues/6043)).

ttoomey committed:
186801632-rhf-refactor

cleanup, add alerting

gigxz committed:
Merge branch 'release-152' into 186801632-rhf-refactor

gigxz committed:
Merge branch 'release-152' into 186801632-rhf-refactor

gigxz commented on src/modules/form/hooks/rhf/useEnrichedFormData.tsx:
```diff
@@ -0,0 +1,95 @@
+  });
+
+  const [defaultValues] = useState<DefaultValues<T>>(() => {
+    const newValues: Record<string, any> = {
+      ...getInitialValues(definition, localConstants),
+      ...cloneDeep(initialValues),
+    };
```
Things seem to be working well during my testing, so I'm inclined to leave this as proposed, but I'm not totally understanding why we added this here. 

Previously the parent component of the Form was responsible for generating the initial values (Eg AssessmentForm)-- and they still do-- but now we also re-generate the initial values values and then overlay the parent-provided initialValues on top of it. 

It does make more sense to me for the initial value calculation to live here. Just calling out that the calculation occurs elsewhere too.

ttoomey replied:
Oh interesting I didn't realize the caller was also performing this. I just extracted it from `useFormDefinitionHandlers`, definitely open to removing it if it's redundant.

gigxz replied:
Oh okay. Let's leave it because I'm not totally sure if its redundant, maybe it is for assessments but not other form types .

gigxz commented on src/modules/form/components/DynamicForm.tsx:
```diff
@@ -167,132 +98,89 @@ const DynamicForm = forwardRef(
+  return (
+    <DynamicFormWithHandlers
+      defaultValues={defaultValues}
+      {...props}
```
`initialValues` is no longer used by any child components now that we are using "defaultValues". Suggest not passing it on? (to avoid confusion of why it's there in the first place)


```diff
   onDirty?: (value: boolean) => void;
-  initialValues?: InitialValues;
   localConstants?: LocalConstants;
   variant?: 'standard' | 'without_top_level_cards';
   errorFilter?: ErrorFilterFn;
@@ -65,7 +64,6 @@ export const DynamicFormWithoutHandlers = forwardRef<
       onSubmit,
       onSaveDraft,
       onDirty,
-      initialValues,
       errors: errorState,
       localConstants,
       handlers,
@@ -158,12 +156,12 @@ const DynamicFormWithHandlers = forwardRef<
   );
 });

-// load remote data (picklists) to augment form data
-
+// load remote data (picklists) to augment form data (initialValues => defaultValues)
 const DynamicFormEnrichedDataLoader = forwardRef<
   DynamicFormRef,
-  DynamicFormProps
+  DynamicFormProps & { initialValues?: InitialValues }
 >((props, ref) => {
+  const { initialValues, ...rest } = props;
   const { defaultValues, loading } = useEnrichedFormData({
     pickListArgs: props.pickListArgs,
     definition: props.definition,
@@ -177,7 +175,7 @@ const DynamicFormEnrichedDataLoader = forwardRef<
   return (
     <DynamicFormWithHandlers
       defaultValues={defaultValues}
-      {...props}
+      {...rest}
       ref={ref}
```

ttoomey replied:
good point, I'll remove it

gigxz reviewed (approved):
this looks like a huge improvement. thanks for all your work @ttoomey. a couple comments below, but overall I say it's ready to move to QA.


system test run against 152: https://github.com/greenriver/hmis-warehouse/pull/5097

gigxz committed:
Merge branch 'release-152' into 186801632-rhf-refactor

ttoomey committed:
186801632-rhf-refactor

pr feedback

gigxz reviewed (approved):
![yay](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExZ29zZXB2eWw1Z3puM3VpNHM0dnR0eXZseXFwa2swNWg2OGE2cXY3MSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5SzzikAmQu6MnDfoml/giphy.gif)

ttoomey committed:
Merge branch 'release-153' into 186801632-rhf-refactor
